### PR TITLE
Adds Compare And Swap to Hyperbee

### PR DESCRIPTION
**Pull request by utanapishtim**
Adds the ability to compare and swap on a key in a Hyperbee. In particular, if we have:

* a hyperbee: `bee`
* a key: `key`
* a value we want to set: `value`
* the current value in `bee` at `key`: `current`
* a value we want to compare against `current`: `test`
* and a predicate that will compare `test` and the current value in `bee` at `key`: `compare`

Then `Hyperbee.cas()` compares `current` and `test` through `compare` and:

if `true`: sets `key` at `value` in `bee` and returns `current`
if `false`: return `null`

The default `compare` function encodes `test` through `bee.valueEncoding` and compares against the encoded value of `current`.

### CODE REVIEW

**Comment by utanapishtim** [comment]
@mafintosh do I need to grab the lock for this like in _put?

File: index.js [hunk 1]
```diff
@@ -556,7 +563,8 @@ class Batch {
         c = Buffer.compare(key, await node.getKey(mid))
 
         if (c === 0) {
-          return (await this.getBlock(node.keys[mid].seq)).final()
+          const block = await this.getBlock(node.keys[mid].seq)
+          return (raw) ? block : block.final()
```
Comment (mafintosh) [inline index.js:567]: The get changes seem unrelated?

Comment (utanapishtim) [reply index.js]: removed

File: README.md [hunk 2]
```diff
@@ -61,37 +61,99 @@ Options include:
...
+Options include:
+
+```js
+{
+  cas (prev, next) { return true }
+}
+```
+
+##### Compare And Swap (cas)
+You can pass a `cas` function as an option to `del` that controls whether the `del` succeeds. Given `bee.del(key, { cas })`, `cas` is passed the current block (i.e. `{ seq, key, value }`) in `bee` at `key` and the next _tentative_ block, in this case the tombstone value `{ seq, key, value: null }`. Then `del` succeeds iff `cas` returns `true` and fails otherwise.
```
Comment (mafintosh) [inline README.md:110]: typo in iff, same above

File: README.md [hunk 3]
```diff
@@ -61,37 +61,99 @@ Options include:
...
+const db = new Hyperbee(feed, { valueEncoding: 'utf8', valueEncoding: 'utf8' })
+await db.put('key', 'value')
+console.log(await db.get('key')) // { seq: 1, key: 'key', value: 'value' }
+await db.del('key', { cas })
+console.log(await db.get('key')) // { seq: 1, key: 'key', value: 'value' }
+await db.put('key', 'value*', { cas })
+console.log(await db.get('key')) // { seq: 2, key: 'key', value: 'value*' }
+await db.del('key', { cas })
+console.log(await db.get('key')) // null
+```
```
Comment (mafintosh) [inline README.md:123]: @andrewosh wanna quickly review the docs here?

**Comment by mafintosh** [comment]
1.9.0
